### PR TITLE
Block Hooks: Add field to block registration, REST API

### PR DESCRIPTION
**Pull request by ockham**
In order to implement Block Hooks (see [`#59313`](https://core.trac.wordpress.org/ticket/59313)), we need to add a new `block_hooks` field to the `WP_Block_Type` class, as well as to block registration related functions, the block type REST API controller, etc.

Based on #5158.

See #3942 for precedent of adding a new field like this.

Trac ticket: https://core.trac.wordpress.org/ticket/59346

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (ockham) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:721]: I've updated this based on https://github.com/WordPress/gutenberg/blob/8c5dbabe5361e94f18e4a55070db9a54d3c59d0b/lib/compat/wordpress-6.4/block-hooks.php#L341-L351. Hope that makes sense!

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php]: Can't we just use `properties`. Like this 

https://github.com/WordPress/wordpress-develop/blob/6f74f616b8973fdef00e81e0984d33c31b229371/src/wp-includes/rest-api/endpoints/class-wp-rest-widgets-controller.php#L836-L858

**Comment by ockham** [comment]
Rebased.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:721]: I don't think that is correct and should changed in gutenberg. This pattern is not needed and make it extremely hard to read.

File: tests/phpunit/tests/rest-api/rest-block-type-controller.php
Comment (ockham) [inline tests/phpunit/tests/rest-api/rest-block-type-controller.php:705]: @gziolo BTW is this list missing `selectors`?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php]: I don't think so, but maybe I'm mistaken? 🤔 

The kind of shape we support is

```json
"blockHooks": {
	"core/comment-template": "lastChild",
	"core/postContent": "after",
}
```

So the keys can be pretty much any block type, whereas the values can only be one of `before`, `after`, `firstChild`, and `lastChild`.

It was my understanding that [`properties`](https://json-schema.org/understanding-json-schema/reference/object.html#properties) is used to define an object with a set of fixed _keys_ -- which we don't really have here. I thus chose [`patternProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties) to allow any block type -- i.e. any alphanumeric string -- as key, while limiting the values to the aforementioned set of possible relative positions.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:721]: See [my other comment](https://github.com/WordPress/wordpress-develop/pull/5203#discussion_r1325709612) 😄

File: tests/phpunit/tests/rest-api/rest-block-type-controller.php
Comment (gziolo) [reply tests/phpunit/tests/rest-api/rest-block-type-controller.php:705]: Yes, it's likely an issue here.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (gziolo) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php]: The way the schema is defined for `block.json` is at https://github.com/WordPress/gutenberg/blob/40c9f17199b067c75d58f1d918fc26c31b08897f/schemas/json/block.json#L709-L718:

```json
"blockHooks": {
	"description": "Block Hooks allow a block to automatically insert itself next to all instances of a given block type.\n\nSee the Block Hooks documentation at https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/#block-hooks-optional for more details.",
	"type": "object",
	"patternProperties": {
		"^[a-z][a-z0-9-]*/[a-z][a-z0-9-]*$": {
			"enum": [ "before", "after", "firstChild", "lastChild" ]
		}
	},
	"additionalProperties": false
},
```

Comment (gziolo) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php]: I see now the comment from @ockham https://github.com/WordPress/wordpress-develop/pull/5203#discussion_r1324725831 that uses:

```php
array(
	'description'       => __( 'This block is automatically inserted near any occurence of the block types used as keys of this map, into a relative position given by the corresponding value.', 'gutenberg' ),
	'patternProperties' => array(
		'^[a-zA-Z0-9-]+/[a-zA-Z0-9-]+$' => array(
			'type' => 'string',
			'enum' => array( 'before', 'after', 'first_child', 'last_child' ),
		),
	),
)
```

It's very close. We should probably unify some aspects. As far as I follow the spec, when using `enum`, you don't need to define the type, but maybe it's required here.

File: tests/phpunit/tests/rest-api/rest-block-type-controller.php
Comment (gziolo) [reply tests/phpunit/tests/rest-api/rest-block-type-controller.php:705]: Although, it's present in the response:

https://github.com/WordPress/wordpress-develop/blob/f3a04cb121bcd4e4df61aeb08837a3182c54cb28/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L525-L532

I see it only in the dev docs:

https://developer.wordpress.org/rest-api/reference/block-types/

Comment (ockham) [reply tests/phpunit/tests/rest-api/rest-block-type-controller.php:705]: I think we should file a separate PR for this, mostly for visibility.

Comment (gziolo) [reply tests/phpunit/tests/rest-api/rest-block-type-controller.php:705]: I can locate the old ticket and commit with a reference to it as a follow-up in the next hour.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php]: [Right:](https://json-schema.org/understanding-json-schema/reference/generic.html#enumerated-values)

> You can use enum even without a type, to accept values of different types. Let’s extend the example to use null to indicate “off”, and also add 42, just for fun.
>
>
> ```json
> {
>  "enum": ["red", "amber", "green", null, 42]
> }
> ```

So I don't think the `type` is required but at least permissible.

File: tests/phpunit/tests/rest-api/rest-block-type-controller.php
Comment (ockham) [reply tests/phpunit/tests/rest-api/rest-block-type-controller.php:705]: @gziolo That'd be great, thank you! I think it's https://core.trac.wordpress.org/ticket/57585 and https://core.trac.wordpress.org/changeset/55673 😊

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php]: As for the `description`, I think it's okay if they're different: After all, in the REST API, we kind of inform the consumer what the field is about:

> This block is automatically inserted near any occurence of the block types used as keys of this map, into a relative position given by the corresponding value.

...whereas in the `block.json` schema, we inform people who write their own block what the `block_hooks` field is used for:

> Block Hooks allow a block to automatically insert itself next to all instances of a given block type.\n\nSee the Block Hooks documentation at https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/#block-hooks-optional for more details.

I think what we have now kind of covers each target audience just fine 👍

**Review by gziolo — APPROVED** [review]
That should cover adding the `block_hooks` field to `WP_Block_Type` and expose it in the REST API endpoint for block types.

Let's continue discussing the shape for the schema after landing initial changes. It's a more complex case as pointed out by @ockham, because the key is the block name that can be anything that matches the pattern for the block name. If that helps we could include the pattern also for the `name` field:

https://github.com/WordPress/wordpress-develop/blob/72de12010b0bd67f814e699705e709f3a125351e/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L478-L484

I see there is custom logic for mapping `blockHooks` to `block_hooks`. I'd like to work on a follow-up to cover that part with unit tests. `before` case is covered indirectly by other values still could use some testing.

One last bit of feedback is for the `_doing_it_wrong` logic. I like that there is additional security enforced. I'm wondering if we should move it to the `WP_Block_Type` class though, to cover also the case when someones registers the block with:

```php
register_block_type( 'my-plugin/my-block', array(
    'block_hooks' => array(
        'my-plugin/my-block': 'first_child',
    ),
) );
```

It also reminds me that in general there isn't that much sanitization present on the code for settings provided during block registration which I find a bit unfortunate.

---

There is nothing blocking this PR, and we can iterate on the items I listed if they make sense.

**Comment by ockham** [comment]
Thank you @gziolo!

> [...] the key is the block name that can be anything that matches the pattern for the block name. If that helps we could include the pattern also for the `name` field:
> 
> https://github.com/WordPress/wordpress-develop/blob/72de12010b0bd67f814e699705e709f3a125351e/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L478-L484

That reminded me of something else; maybe we can cross-reference the `name` field in the `block_hooks` schema definition? Would make it more semantic -- we'd basically state, "The keys in `block_hooks` are block `name`s.".
 
> I see there is custom logic for mapping `blockHooks` to `block_hooks`. I'd like to work on a follow-up to cover that part with unit tests. `before` case is covered indirectly by other values still could use some testing.

Ah yeah, that's a fair point; especially for the camelCase -> snake_case mappings.

> One last bit of feedback is for the `_doing_it_wrong` logic. I like that there is additional security enforced. I'm wondering if we should move it to the `WP_Block_Type` class though, to cover also the case when someones registers the block with:
> 
> ```
> register_block_type( 'my-plugin/my-block', array(
>     'block_hooks' => array(
>         'my-plugin/my-block': 'first_child',
>     ),
> ) );
> ```

FWIW, `register_block_type` actually calls `register_block_type_from_metadata` for block registration, so your example should already be covered: https://github.com/WordPress/wordpress-develop/blob/e005108fe18e6278f4e6dd0fec030e503fe0ff46/src/wp-includes/blocks.php#L584-L590

(We might want to add some unit test for this too, BTW 🤔 )

> It also reminds me that in genera; there isn't that much sanitization present on the code for settings provided during block registration which I find a bit unforunate.

Yeah, good point!

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (gziolo) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php]: Yes, the description can differ, but for example, pattern matching for the block name should always be the same.

By the way, it should be as simple as adding `pattern` to the `name` field:

https://json-schema.org/understanding-json-schema/reference/regular_expressions.html#example

**Comment by ockham** [comment]
Committed to Core in https://core.trac.wordpress.org/changeset/56587.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:713]: @ockham How something like this.

```
$properties = array();
$registry   = WP_Block_Type_Registry::get_instance();
foreach( $registry->get_all_registered() as $block ){
   $properties[ $block ] = array(
       'type' => 'string',
       'enum' => array( 'before', 'after', 'first_child', 'last_child' ),
   );
}

File: tests/phpunit/tests/rest-api/rest-block-type-controller.php
Comment (gziolo) [reply tests/phpunit/tests/rest-api/rest-block-type-controller.php:705]: It was addressed in https://core.trac.wordpress.org/changeset/56588.

File: src/wp-includes/blocks.php
Comment (Soean) [inline src/wp-includes/blocks.php:536]: We should remove the 'gutenberg' text domain.

```suggestion
					__( 'Cannot hook block to itself.' ),
```

Comment (Soean) [reply src/wp-includes/blocks.php:536]:
```suggestion
					__( 'Cannot hook block to itself.' ),
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
Comment (gziolo) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:713]: Noting that there might be hundreds of registered blocks, but the block author might still list the block type name that isn't present on the site and we would consider it valid. What's the reasoning behind listing all available blocks?

File: src/wp-includes/blocks.php
Comment (gziolo) [reply src/wp-includes/blocks.php:536]: I'm on it 👍🏻

Comment (gziolo) [reply src/wp-includes/blocks.php:536]: https://github.com/WordPress/wordpress-develop/pull/5218 - PR is ready.

**Comment by gziolo** [comment]
In https://github.com/WordPress/wordpress-develop/pull/5218 I added some refactorings discussed above:
- https://github.com/WordPress/wordpress-develop/pull/5203#discussion_r1326102815 - `gutenberg` text domain remove
- There is custom logic for mapping `blockHooks` to `block_hooks` when handling metadata. I covered that part in unit tests.
- One last bit of feedback was for the _doing_it_wrong logic. I moved the check to the `WP_Block_Type` class to cover also the case when someone registers the block with a regular `register_block_type` call without using metadata file.

File: src/wp-includes/blocks.php
Comment (gziolo) [reply src/wp-includes/blocks.php:536]: I committed with https://core.trac.wordpress.org/changeset/56592 as there is still ongoing discussion in the PR.

**Pull request closed without merging**
